@CacheResult on legacy @RegisterRestClient throws CCE on 3.4.2

### Describe the bug

Using `@CacheResult` with legacy `@RegisterRestClient` throws `ClassCastException` on 3.4.2


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

Using the following dependencies:
* quarkus-cache
* quarkus-resteasy
* quarkus-rest-client
* quarkus-rest-client-jackson

Given the below example service and resource classes:
```java
package org.acme.cache;

import org.eclipse.microprofile.rest.client.inject.RestClient;
import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;

@Path("/ip")
public class IpResource {
    @Inject
    @RestClient
    IpService service;

    @GET
    public String getIp() {
        return service.getIp();
    }
}
```

```java
package org.acme.cache;

import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
import io.quarkus.cache.CacheResult;

@Path("/")
@RegisterRestClient(configKey="ip")
public interface IpService {
  @GET
  @CacheResult(cacheName = "ip")
  String getIp();
}
```

and `ip/mp-rest/url=https://api.ipify.org/` entry in `application.properties`.

On Quarkus 3.3.3, it works:
> $ curl localhost:8080/ip
90.xxx.xx.xxx

On Quarkus 3.4.2:
> $ curl localhost:8080/ip
{"details":"Error id 778dc968-4c07-44eb-b8ab-983596bce955-1, org.jboss.resteasy.spi.UnhandledException: java.lang.ClassCastException: class jdk.proxy3.$Proxy65 cannot be cast to class io.quarkus.arc.AbstractAnnotationLiteral .....


Exception stack trace:
```
Caused by: java.lang.ClassCastException: class jdk.proxy3.$Proxy65 cannot be cast to class io.quarkus.arc.AbstractAnnotationLiteral (jdk.proxy3.$Proxy65 is in module jdk.proxy3 of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @77e9807f; io.quarkus.arc.AbstractAnnotationLiteral is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @77e9807f)
        at io.quarkus.cache.runtime.CacheInterceptor.getArcCacheInterceptionContext(CacheInterceptor.java:76)
        at io.quarkus.cache.runtime.CacheInterceptor.getInterceptionContext(CacheInterceptor.java:56)
        at io.quarkus.cache.runtime.CacheResultInterceptor.intercept(CacheResultInterceptor.java:38)
``` 

Commenting out the `@CacheResult` lines makes the rest client work on 3.4.2 as well.



### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "17.0.8" 2023-07-18 OpenJDK Runtime Environment Temurin-17.0.8+7 (build 17.0.8+7) OpenJDK 64-Bit Server VM Temurin-17.0.8+7 (build 17.0.8+7, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.4.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546)

### Additional information

_No response_

--- a/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java
+++ b/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java
@@ -71,9 +71,10 @@ public abstract class CacheInterceptor {
     private <T> Optional<CacheInterceptionContext<T>> getArcCacheInterceptionContext(
             InvocationContext invocationContext, Class<T> interceptorBindingClass) {
         Set<AbstractAnnotationLiteral> bindings = InterceptorBindings.getInterceptorBindingLiterals(invocationContext);
-        if (bindings == null) {
-            LOGGER.trace("Interceptor bindings not found in ArC");
-            // This should only happen when the interception is not managed by Arc.
+        if (bindings == null || bindings.isEmpty() || !(bindings.iterator().next() instanceof AbstractAnnotationLiteral)) {
+            // this should only happen when the interception is not managed by ArC
+            // a non-`AbstractAnnotationLiteral` can come from RESTEasy Classic's `QuarkusInvocationContextImpl`
+            LOGGER.trace("Interceptor bindings not found in ArC or not created by ArC");
             return Optional.empty();
         }
         List<T> interceptorBindings = new ArrayList<>();

